### ISSUE TITLE
Support Cython-based async functions

### ISSUE THREAD

**Issue by touilleMan** [python-trio/trio#550]
The title say it all, some projects re-implement their own iscoroutine function by looking for `CO_COROUTINE` and `CO_ITERABLE_COROUTINE` flags ([for instance](https://github.com/ssanderson/interface/pull/17/files#diff-62977b350e62b018986bf62bcbace77aR63)) to address this...

This is more of a CPython issue of course, but this currently means the coroutine generated by cython are not compatible with trio:
```python
import trio

async def trio_main():
    print('hello...')
    await trio.sleep(1)
    print(' world !')

def main():
    trio.run(trio_main)

if __name__ == '__main__':
    main()
```

```
$ python cython_test.py 
hello...
 world !
$ cythonize -a -i cython_test.py 
Compiling /home/emmanuel/projects/parsec-lrw/cython_test.py because it changed.
[1/1] Cythonizing /home/emmanuel/projects/parsec-lrw/cython_test.py
running build_ext
building 'cython_test' extension
creating /home/emmanuel/projects/parsec-lrw/tmp0p_t5pyt/home
creating /home/emmanuel/projects/parsec-lrw/tmp0p_t5pyt/home/emmanuel
creating /home/emmanuel/projects/parsec-lrw/tmp0p_t5pyt/home/emmanuel/projects
creating /home/emmanuel/projects/parsec-lrw/tmp0p_t5pyt/home/emmanuel/projects/parsec-lrw
gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I/home/emmanuel/projects/cpython/build-install/include/python3.6m -c /home/emmanuel/projects/parsec-lrw/cython_test.c -o /home/emmanuel/projects/parsec-lrw/tmp0p_t5pyt/home/emmanuel/projects/parsec-lrw/cython_test.o
gcc -pthread -shared /home/emmanuel/projects/parsec-lrw/tmp0p_t5pyt/home/emmanuel/projects/parsec-lrw/cython_test.o -o /home/emmanuel/projects/parsec-lrw/cython_test.cpython-36m-x86_64-linux-gnu.so
$ python 
Python 3.6.3 (v3.6.3:2c5fed8, Feb  6 2018, 17:05:14) 
[GCC 5.4.1 20160904] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import cython_test
>>> cython_test.main()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "cython_test.py", line 10, in cython_test.main
    trio.run(trio_main)
  File "/home/emmanuel/projects/parsec-lrw/venv/lib/python3.6/site-packages/trio/_core/_run.py", line 1238, in run
    return result.unwrap()
  File "/home/emmanuel/projects/parsec-lrw/venv/lib/python3.6/site-packages/trio/_core/_result.py", line 119, in unwrap
    raise self.error
  File "/home/emmanuel/projects/parsec-lrw/venv/lib/python3.6/site-packages/trio/_core/_run.py", line 1347, in run_impl
    msg = task.context.run(task.coro.send, next_send)
  File "/home/emmanuel/projects/parsec-lrw/venv/lib/python3.6/site-packages/contextvars/__init__.py", line 38, in run
    return callable(*args, **kwargs)
  File "/home/emmanuel/projects/parsec-lrw/venv/lib/python3.6/site-packages/trio/_core/_run.py", line 934, in init
    async_fn, args, system_nursery, None
  File "/home/emmanuel/projects/parsec-lrw/venv/lib/python3.6/site-packages/trio/_core/_run.py", line 805, in spawn_impl
    .format(getattr(async_fn, "__qualname__", async_fn))
TypeError: trio expected an async function, but 'trio_main' appears to be synchronous
```

I didn't dig into CPython internal so my understanding on why `inspect.iscoroutine` != `asyncio.iscoroutine` on this is fairly limited, but I guess we should do something on our own to fix this anyway ;-)

**Comment by sorcio** [comment]
Not sure how relevant but Cython changelog for 0.23.0 says:

> When generators are used in a Cython module and the module imports the modules “inspect” and/or “asyncio”, Cython enables interoperability by patching these modules during the import to recognise Cython’s internal generator and coroutine types. This can be disabled by C compiling the module with “-D CYTHON_PATCH_ASYNCIO=0” or “-D CYTHON_PATCH_INSPECT=0”

So maybe the difference is with vanilla vs patched `inspect.iscoroutine()`?

(Did you mean "Cython issue", not "CPython issue", right?)

**Comment by touilleMan** [comment]
> (Did you mean "Cython issue", not "CPython issue", right?)

From my point of view the trouble comes from the fact CPython provide two `iscoroutine` functions giving different results depending on how a coroutine is created.

Cython just choose the "wrong" coroutine style for `inspect.iscoroutine` which is unfortunate for Trio.

> So maybe the difference is with vanilla vs patched inspect.iscoroutine()?

Really interesting point, this seems to indicate Cython patched both inspect and asyncio's iscoroutine, so both should work which is of course no the case... Maybe it's a Cython bug after all ^^

**Comment by njsmith** [comment]
@scoder Do you have any insights here? It sounds like the issue is that `inspect.iscoroutine` is returning `False` for a Cython coroutine – is that expected? Would using `isinstance(obj, abc.Coroutine)` work better than `inspect.iscoroutine(obj)`?

**Comment by scoder** [comment]
> So maybe the difference is with vanilla vs patched inspect.iscoroutine()?

As the docs say, Cython only patches these if the compiled module itself imports them directly. Since the compiled module only imports `trio` and neither `asyncio` nor `inspect`, there is no monkey patching going on here. (Although in a real deployment, other Cython compiled modules might have done the patching for you. That's the usual problem with this kind of monkey patching.)

> inspect.iscoroutine is returning False for a Cython coroutine – is that expected?

Yes. The `inspect` module is based on type checks, not protocols. 

> Would using isinstance(obj, abc.Coroutine) work better than inspect.iscoroutine(obj)?

It's probably what you want. The only reason to require exactly Python coroutine functions is when you depend on their C-level details or other implementation specifics. In all other cases, any object that supports the coroutine protocol should be enough.

**Comment by njsmith** [comment]
> Cython only patches these if the compiled module itself imports them directly

Ah, that is pretty confusing. I feel like I would either never patch, or else always patch `inspect` in any program where cython coroutines can occur. (I believe `asyncio` uses `inspect` these days, and `inspect` is a much lighter import than `asyncio`.) You've obviously thought about this a lot more than me though, so there might be some downside that I'm missing.

> The only reason to require exactly Python coroutine functions is when you depend on their C-level details or other implementation specifics

We don't care about the C-level details, but in general I think we assume that coroutine objects really implement the *full* coroutine interface, including the `cr_*` attributes. Technically the `cr_*` attributes are not part of the `abc.Coroutine` interface. Do Cython coroutines aim to expose the full coroutine interface, or just the `abc.Coroutine` interface?

**Comment by scoder** [comment]
> that is pretty confusing

I prefer the word non-obvious, but yes. Any kind of stdlib monkey patching has the downside of being unexpected for some code. I'd rather not patch at all.

> Do Cython coroutines aim to expose the full coroutine interface

As much as they can. They don't have actual byte code, obviously, nor a frame, but they have a valid `cr_code`, `cr_running` and `cr_await`. I doubt that you need more than that.

**Comment by njsmith** [comment]
Proposed fix in #550. I haven't actually tested it though :-). @touilleMan, would it be easy for you to try it out?

**Comment by kawing-chiu** [comment]
I think I just hit this bug (on cython `0.29.2`, trio `0.9.0`). Trio now raises a `TrioInternalError`. Is there any recommended workaround for now? I've tried importing the inspect, but it does not work.

**Comment by njsmith** [comment]
@kawing-chiu sorry to hear that! If you search the page for the string "workaround", there is one buried in the middle of one of the messages up above. It would be great to hear if it works for you.

**Comment by kawing-chiu** [comment]
Great! I was tired enough to miss that last night. And the workaround does work. Thanks very much!

**Comment by njsmith** [comment]
Looking at this again, I suppose we could, like, automatically apply the workaround when this happens?

**Comment by xueyoo** [comment]
environment:  cython:0.29.23  and trio:0.18.0
issue: 'NoneType' object has no attribute 'f_locals'

fix code:
trio/_core/_run.py
if not hasattr(coro, "cr_frame")
change to
if not hasattr(coro, "cr_frame") or coro.cr_frame is None:

**Comment by scoder** [comment]
@xueyoo, I backported the implementation of `cr_frame` to Cython 0.29.24 that was previously only available in 3.0. Feel free to use 3.0 (alpha), though.

**Comment by pquentin** [comment]
@scoder Thank you! This helped @didimelli in chat: https://gitter.im/python-trio/general?at=60a23cd601edef5b504d292f
